### DISCUSSION TITLE
How to determine if a function has been manually edited?

### DISCUSSION THREAD

**Discussion by alex-bellon** [NationalSecurityAgency/ghidra#4427] [Q&A]
I am currently writing a Ghidra script that will go through functions and mark certain registers as containing parameters for the function. However, I don't want to apply this change to functions where someone has already manually messed with the parameters. 

Right now I was thinking of passing the something with [lower `SourceType` precendence](https://github.com/NationalSecurityAgency/ghidra/blob/b8425e804d5b1c4c8d9d739827ed5260a9c4c9d1/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/model/symbol/SourceType.java#L22-L29) than `USER_DEFINED` into my call to [`replaceParameters()`](https://github.com/NationalSecurityAgency/ghidra/blob/b8425e804d5b1c4c8d9d739827ed5260a9c4c9d1/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/model/listing/Function.java#L379), but I looked through the source for `replaceParameters()` and I'm not sure that would work.

Is there a way to tell if a function (or parameters) have been manually changed?

**Comment by ghidra1** [comment]
There is a separate source type for each parameter, although not the return param, which would reflect any direct change to a parameter - although much depends on how the change was made and the source type specified.  In addition, the Function signature source type reflects changes made to the entire signature.  The source type `USER_DEFINED` is intended to indicate those elements which have been manually changed and not set via import or analysis.
